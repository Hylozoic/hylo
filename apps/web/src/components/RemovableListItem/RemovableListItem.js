import React from 'react'
import { useTranslation } from 'react-i18next'
import { Link } from 'react-router-dom'
import { DEFAULT_AVATAR } from 'store/models/Group'
import RoundImage from 'components/RoundImage'
import classes from './RemovableListItem.module.scss'

export default function RemovableListItem ({ item, removeItem, skipConfirm = false, square, size, confirmMessage, url }) {
  const { t } = useTranslation()
  const remove = () => {
    if (skipConfirm) return removeItem(item.id)

    confirmMessage = confirmMessage || t('Are you sure you want to remove {{item}}?', { item: item.name || item.title })
    if (window.confirm(confirmMessage)) {
      removeItem(item.id)
    }
  }

  const avatar = item.avatarUrl ? <RoundImage url={item.avatarUrl || DEFAULT_AVATAR} medium square={square} size={size} className={classes.avatar} /> : null
  const title = item.name || item.title

  return (
    <div className='bg-card rounded-lg p-2 flex justify-between items-center'>
      <div>
        {url && <Link to={url}>{avatar}</Link>}
        {!url && avatar}

        {url && <Link to={url} className='text-foreground/100'>{title}</Link>}
        {!url && <span>{title}</span>}
      </div>

      {removeItem && <span onClick={remove} className='border-2 border-accent/20 rounded-lg p-1 text-xs text-accent/60 hover:text-accent/100 hover:bg-accent/10 hover:border-accent/40 transition-all cursor-pointer'>{t('Remove')}</span>}
    </div>
  )
}
