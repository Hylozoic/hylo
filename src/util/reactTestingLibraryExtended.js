import React from 'react'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import rootReducer from 'store/reducers'
import createMiddleware from 'store/middleware'

export function generateStore (history, initialState) {
  return createStore(rootReducer, initialState, createMiddleware(history))
}

export const AllTheProviders = (store) => ({ children }) => {
  return (
    <Provider store={store || createStore(() => {})}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
}

// re-export everything

const rtl = require('react-testing-library')
const customRender = (ui, options, providersFunc) =>
  rtl.render(ui, { wrapper: providersFunc, ...options })

module.exports = {
  ...rtl,
  render: customRender
}
