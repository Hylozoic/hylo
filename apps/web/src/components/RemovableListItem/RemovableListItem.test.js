import React from 'react'
import { render, screen, fireEvent } from 'util/testing/reactTestingLibraryExtended'
import RemovableListItem from './RemovableListItem'

describe('RemovableListItem', () => {
  const defaultItem = {
    id: 7,
    name: 'Zeus',
    avatarUrl: 'zeus.png'
  }

  it('renders correctly with URL', () => {
    render(
      <RemovableListItem item={defaultItem} url="/happy/place" removeItem={() => {}} />
    )

    expect(screen.getByText('Zeus')).toBeInTheDocument()
    expect(screen.getByRole('link', { name: 'Zeus' })).toHaveAttribute('href', '/happy/place')
    expect(screen.getByRole('img')).toHaveAttribute('src', 'zeus.png')
    expect(screen.getByText('Remove')).toBeInTheDocument()
  })

  it('does not render as links when no URL specified', () => {
    render(<RemovableListItem item={defaultItem} removeItem={() => {}} />)

    expect(screen.getByText('Zeus')).toBeInTheDocument()
    expect(screen.queryByRole('link')).not.toBeInTheDocument()
    expect(screen.getByRole('img')).toHaveAttribute('src', 'zeus.png')
  })

  it('doesnt render a remove link when removeItem is not provided', () => {
    render(
      <RemovableListItem item={defaultItem} url="/happy/place" />
    )

    expect(screen.getByText('Zeus')).toBeInTheDocument()
    expect(screen.queryByText('Remove')).not.toBeInTheDocument()
  })

  describe('remove item', () => {
    it('calls remove with confirmation', () => {
      const removeItem = jest.fn()
      const confirmMessage = 'Are you sure?'

      window.confirm = jest.fn(() => true)

      render(
        <RemovableListItem
          item={defaultItem}
          url="/happy/place"
          removeItem={removeItem}
          confirmMessage={confirmMessage}
        />
      )

      fireEvent.click(screen.getByText('Remove'))

      expect(window.confirm).toHaveBeenCalledWith(confirmMessage)
      expect(removeItem).toHaveBeenCalledWith(defaultItem.id)
    })

    it('skips confirm when skipConfirm is true', () => {
      const removeItem = jest.fn()

      window.confirm = jest.fn()

      render(
        <RemovableListItem
          item={defaultItem}
          skipConfirm={true}
          url="/happy/place"
          removeItem={removeItem}
        />
      )

      fireEvent.click(screen.getByText('Remove'))

      expect(window.confirm).not.toHaveBeenCalled()
      expect(removeItem).toHaveBeenCalledWith(defaultItem.id)
    })
  })
})
