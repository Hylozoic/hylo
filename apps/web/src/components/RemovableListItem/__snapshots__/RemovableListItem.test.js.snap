// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemovableListItem does not render as links when no URL specified 1`] = `
<div
  data-stylename="item"
>
  <RoundImage
    data-stylename="avatar"
    medium={true}
    url="zeus.png"
  />
  <span>
    Zeus
  </span>
  <span
    className="remove-button"
    data-stylename="remove-button"
    onClick={[Function]}
  >
    Remove
  </span>
</div>
`;
